src/storage/data_table.cpp

```diff
@@ -180,24 +225,65 @@ bool DataTable::ScanBaseTable(Transaction &transaction, DataChunk &result, Table
 		current_row += STANDARD_VECTOR_SIZE;
 		return true;
 	}
-
-	for (idx_t i = 0; i < state.column_ids.size(); i++) {
-		auto column = state.column_ids[i];
-		if (column == COLUMN_IDENTIFIER_ROW_ID) {
-			// scan row id
-			assert(result.data[i].type == ROW_TYPE);
-			result.data[i].Sequence(base_row + current_row, 1);
+	//! If we have filters
+	if (!table_filters.empty() && !apply_filter) {
+		SelectionVector sel;
+		idx_t approved_tuple_count = count;
+		if (count != max_count) {
+			sel = valid_sel;
 		} else {
-			// scan actual base column
-			columns[column].Scan(transaction, state.column_scans[i], result.data[i]);
+			sel = FlatVector::IncrementalSelectionVector;
+		}
+		//! First, we scan the columns with filters, fetch their data and generate a selection vector.
+		for (auto &table_filter : table_filters) {
+			apply_filter = columns[table_filter.first].Select(transaction, state.column_scans[table_filter.first],
+			                                                  result.data[table_filter.first], sel,
+			                                                  approved_tuple_count, table_filter.second);
+			if (apply_filter) {
+				break;
+			}
+		}
+		if (!apply_filter) {
+			result.Slice(sel, approved_tuple_count);
+			//! Now we use the selection vector to fetch data for the other columns.
+			for (idx_t i = 0; i < state.column_ids.size(); i++) {
+				if (table_filters.find(i) == table_filters.end()) {
+					auto column = state.column_ids[i];
+					if (column == COLUMN_IDENTIFIER_ROW_ID) {
+						assert(result.data[i].type == TypeId::INT64);
+						result.data[i].vector_type = VectorType::FLAT_VECTOR;
+						auto result_data = (int64_t *)FlatVector::GetData(result.data[i]);
+						for (size_t sel_idx = 0; sel_idx < approved_tuple_count; sel_idx++) {
+							result_data[sel_idx] = base_row + current_row + sel.get_index(sel_idx);
+						}
+					} else {
+						columns[column].FilterScan(transaction, state.column_scans[i], result.data[i], sel,
+						                           approved_tuple_count);
+					}
+				}
+			}
+			result.SetCardinality(approved_tuple_count);
 		}
 	}
-	if (count == max_count) {
-		// no deleted tuples
-		result.SetCardinality(count);
-	} else {
-		// deleted tuples
-		result.Slice(valid_sel, count);
+
+	if (apply_filter || table_filters.empty()) {
+		for (idx_t i = 0; i < state.column_ids.size(); i++) {
+			auto column = state.column_ids[i];
+			if (column == COLUMN_IDENTIFIER_ROW_ID) {
+				// scan row id
+				assert(result.data[i].type == ROW_TYPE);
+				result.data[i].Sequence(base_row + current_row, 1);
+			} else {
+				columns[column].Scan(transaction, state.column_scans[i], result.data[i]);
```

Comment:
Since you implemented the ColumnData::FilterScan, this can also be used here if there are deleted tuples. That should be more efficient than fetching the entire vector and then slicing it.